### PR TITLE
issue: 865066 Merge experimental branch into master branch

### PR DESCRIPTION
**Pull request by OphirMunk**
The merged experimental branch adds a new VMA API based on direct CQ polling
(named "vmapoll") as a fast alternative to legacy recieve socket API.

VMA can run either in legacy socket API or in vmapoll API.

In order to build VMA to run in legacy socket API please execute
$ configure
$ make

In order to build VMA to run in vmapoll API please execute
$ configure --enable-vmapoll
$ make

A new file vmapoll_extra.h was added to define vmapoll API.

Signed-off-by: Ophir Munk <ophirmu@mellanox.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/691/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/691/ for details (Mellanox internal link).

File: configure.ac [hunk 1]
```diff
@@ -225,6 +282,217 @@ else
...
+    [],
+    [with_rdtsc_receivefrom_to_sendto=no]
+)
+
+AS_IF([test "x$with_rdtsc_receivefrom_to_sendto" == xyes],
+	[AC_DEFINE([RDTSC_MEASURE], 1, [Define to 1 to enable rdtsc measurements.])]
+	[AC_DEFINE([RDTSC_MEASURE_RECEIVEFROM_TO_SENDTO], 1, [Define to 1 to enable rdtsc measurement of receivefrom to sendto.])]
+	[AC_MSG_RESULT([yes])],
+	[AC_MSG_RESULT([no])]
+)
```
Comment (igor-ivanov) [inline configure.ac]: It looks as this code related measurements interferes with https://github.com/Mellanox/libvma/pull/194 and probably should be done separate pull request.

**Comment by OphirMunk** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/698/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/699/ for details (Mellanox internal link).

**Comment by igor-ivanov** [comment]
@OphirMunk Could you please consider splitting this huge commit into smaller, more manageable parts ~10.

I see at least following reasons to do it.
* It is a big challenge to review large commit at once (It could be reviewed, accepted step-by-step).
* It is harder to track the order that changes were made, and hence to understand the cause of problems in the future.
* It is problematic to link this changes with related tickets. In many cases tickets are more informative the commit message.

How it can be done:
* use existing commits in feature branch, combine them in logical units to have final set of reasonable commits
* separate basing on functionality for example configure/make, tests, general improvement, changes related this capability etc

Let think how can we understand a logic of such changes, motivation etc after feature branch dropping (In current VMA workflow we have mainstream branch as master and could have temporary feature development branches until it is pushed to mainstream branch).

Also I would like to propose adding a list of PRs that these changes are covered into this pull request description and update commit message with list of issues.

File: src/vma/proto/mem_buf_desc.h [hunk 2]
```diff
@@ -128,6 +134,9 @@ class mem_buf_desc_t {
 			};
 
 			void* 		context;
+#ifdef DEFINED_VMAPOLL 
+			bool 		vma_polled;
+#endif // DEFINED_VMAPOLL
```
Comment (igor-ivanov) [inline src/vma/proto/mem_buf_desc.h]: `vma_polled` is vma_poll specific variable. It means nothing for legacy mode. So it can be visible for all configurations and used in different parts of code to differ modes w/o using a lot of `DEFINED_VMAPOLL`
Please consider enable one for legacy and vma_poll and use this to reduce `DEFINED_VMAPOLL`

File: configure.ac [hunk 3]
```diff
@@ -168,6 +169,62 @@ AS_IF([test "x$enable_exp_cq" == xyes],
...
 
+#
+# VMAPOLL
+#
+AC_ARG_ENABLE(
+    [vmapoll],
+    AC_HELP_STRING(
+        [--enable-vmapoll],
+        [Enable alternative API to standard sockets API for receiving packets (default=no)]))
+AC_MSG_CHECKING(
```
Comment (OphirMunk) [inline configure.ac]: Detailed explanation should be added to User Manual. Configure help is just 1 summary line.

File: configure.ac [hunk 4]
```diff
@@ -168,6 +169,62 @@ AS_IF([test "x$enable_exp_cq" == xyes],
...
+        [Minimize logging output (default=no)]))
+AC_MSG_CHECKING(
+    [if minimized logging output is enabled])
+if test "x$enable_opt_log" = "xyes"; then
+    CPPFLAGS="$CPPFLAGS -DVMA_OPTIMIZE_LOG -DNDEBUG"
+    AC_MSG_RESULT([yes])
+else
+    AC_MSG_RESULT([no])
+fi
+
```
Comment (OphirMunk) [inline configure.ac]: Following our mutual review: comment relates to line 225

Comment (OphirMunk) [inline configure.ac]: logging options are now duplicated in config/m4/opt.m4 and configure.ac files. Need to choose either one

File: configure.ac [hunk 5]
```diff
@@ -168,6 +169,62 @@ AS_IF([test "x$enable_exp_cq" == xyes],
...
+    if test "x$enable_vmapoll" = xyes; then
+        AC_DEFINE([DEFINED_NO_THREAD_LOCK], 1, [Define to 1 to disable thread locking])
+        AC_MSG_RESULT([no])
+    else
+        AC_MSG_RESULT([yes (vmapoll is not enabled)])
+    fi
+else
+    AC_MSG_RESULT([yes])
+fi
+
```
Comment (OphirMunk) [inline configure.ac]: The merge reflects the current functionality of experimental_256B branch
Improvements will be implemented after the merge.

File: configure.ac [hunk 6]
```diff
@@ -450,6 +718,25 @@ AC_TRY_LINK(
...
+  access = access;
+],
+[
+  AC_MSG_RESULT([yes])
+  AC_DEFINE(DEFINED_MLX5_HW_ETH_WQE_HEADER, 1, [Define to 1 if MLX5_ETH_WQE_L3_CSUM is defined])
+],
+[
+  AC_MSG_RESULT([no])
+])
+
```
Comment (OphirMunk) [inline configure.ac]: Good point. Should be addressed following the merge.

File: src/vma/dev/net_device_table_mgr.cpp [hunk 7]
```diff
@@ -708,7 +713,13 @@ void net_device_table_mgr::handle_timer_expired(void* user_data)
 	int timer_type = (uint64_t)user_data;
 	switch (timer_type) {
 	case RING_PROGRESS_ENGINE_TIMER:
+#if 0 /* TODO: see explanation */
+		/* Do not call draining RX logic from internal thread for vma_poll mode
```
Comment (OphirMunk) [inline src/vma/dev/net_device_table_mgr.cpp]: Fixed

File: src/vma/hw/mlx5/wqe.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef WQE_H
+#define WQE_H
+
+#define MLX5_ETH_INLINE_HEADER_SIZE 16
+
```
Comment (OphirMunk) [inline src/vma/hw/mlx5/wqe.h]: Good point. Maybe should consider:
```
#ifndef MLX5_ETH_INLINE_HEADER_SIZE
#define MLX5_ETH_INLINE_HEADER_SIZE 16
#endif // MLX5_ETH_INLINE_HEADER_SIZE
```

File: src/vma/sock/sockinfo_udp.cpp [hunk 9]
```diff
@@ -1146,6 +1155,15 @@ int sockinfo_udp::getsockopt(int __level, int __optname, void *__optval, socklen
...
 
+#ifdef DEFINED_VMAPOLL
+	/* Process VMA specific options only at the moment
+	 * VMA option does not require additional processing after return
+	 */
+	if (0 == sockinfo::getsockopt(__level, __optname, __optval, __optlen)) {
+		return 0;
+	}
+#endif // DEFINED_VMAPOLL	
+
```
Comment (OphirMunk) [inline src/vma/sock/sockinfo_udp.cpp]: Agreed (may implement after the merge)

File: src/vma/dev/ring.cpp [hunk 10]
```diff
@@ -35,4 +35,21 @@
...
+
+ring::~ring()
+{
+#ifdef DEFINED_VMAPOLL
+	struct ring_ec *ec = get_ec();
+
+	while (ec) {
+		clear_ec(ec);
+		ec = get_ec();
+	}
```
Comment (igor-ivanov) [inline src/vma/dev/ring.cpp]: There is https://github.com/Mellanox/libvma/pull/253 that is not part of this change.

File: build/libvma.spec.in [hunk 11]
```diff
@@ -79,6 +79,7 @@ rm -f $RPM_BUILD_ROOT%{_libdir}/*.la
 
 install -m 755 tests/vma_perf_envelope/vma_perf_envelope.sh $RPM_BUILD_ROOT/%{vma_datadir}/scripts/vma_perf_envelope.sh
 install -m 644 src/vma/vma_extra.h $RPM_BUILD_ROOT/%{include_dir}/vma_extra.h
+install -m 644 src/vma/vmapoll_extra.h $RPM_BUILD_ROOT/%{include_dir}/vmapoll_extra.h
```
Comment (igor-ivanov) [inline build/libvma.spec.in]: vmapoll_extra.h is not installed are part of `make install`

**Comment by OphirMunk** [comment]
Merged #262
